### ISSUE TITLE
regarding replicating your baseline results

### ISSUE THREAD

**Issue by kevinghst** [mila-iqia/spr#24]
Hi,

thanks so much for making the codebase available!

I have some questions regarding your baseline results.

First of all, is the following command the correct one to reproduce your baseline result (no SPR, no augmentation), presumably given by the first row of Table 4 (Rainbow controlled, no aug - 0.24 Human Normalized Score) ?

`python -m scripts.run --public --game boxing --augmentation none --target-augmentation 0 --spr 0 --dropout 0.5 
`

If it is, then I have some concerns regarding the results. Here are what I reproduced for baseline, SPR w/o aug, and SPR:

| Game  | DER (from paper) |Baseline | SPR w/o aug | SPR |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Alien  |739| 852 (177)  | 891 (255) | 875 (137) |
| Amidar |188| 160 (40)  | 204 (59) | 205 (36) |
| Assault |431| 739 (13) | 705 (64) | 699 (36) |
| Bank Heist |51| 308 (208)  | 401 (303) | 324 (198) |
| Boxing |0.2| 13 (10)  | 16.57 (7.3) | 28.9 (14) |
| Breakout |1.9| 6.8 (5.5)  | 10.8 (8.6) | 20 (9.1) |

**10 seeds for each experiment**

In particular, the baseline results seem to be much higher than expected. First of all, your paper mentions that your baseline is based on DER, but the baseline results I got for the 6 games are much higher than what is reported for DER on Table 1 of your paper, in particular for the latter 3 games. The fact that your control (Rainbow controlled, no aug) is supposed to perform worse than DER (0.24 vs 0.28 HNS given in Table 4) makes this even more surprising.

In addition, table 6 reports the HNS for SPR w/o augmentation to be 0.46. So one would expect on average SPR w/o aug to achieve twice the score as the baseline, which is again contrary to my data.

I guess it's possible that I got "lucky", since even with 10 seeds the variance is high. But before I go crazy confirming my results by running the rest of the games, I just want to make sure I am replicating your experiments correctly.

Thanks!

**Comment by ankeshanand** [comment]
Hi @kevinghst,

Thanks for the detailed report, appreciate you taking the time to look into this! The Rainbow controlled no aug baseline in the paper used these parameters: 
`python -m scripts.run --public --game boxing --augmentation none --target-augmentation 0 --spr 0` 

Couple of points regarding Human Normalized Scores (HNS):
1. HNS is defined as improvement over random `(score - random_agent_score) / (human_score - random_agent_score)`, so you would expect `SPR - random` to be twice `baseline - random`: this isn't the same as `SPR = 2*baseline`.

2. Morevover, you can't meaningfully estimate mean (or median) HNS from only six games.  Variance of scores across games is high and some games count for much more than others (because they're near the midpoint for median or because they have very high scores for mean). We expand on these points in section 4.2, and also why we report not only the mean and median scores but also the boxplot showing the full distribution of scores in Figure 3. 

Are you already using the commands mentioned in the README to run SPR and SPR w/ aug? You can cross-check with Table 1 to see the scores we got for both of them. 

And keep us posted if you decide to do more experiments and don't get close to the reported numbers. We will be happy to dig in more and help narrow down the source of discrepency.

**Comment by kevinghst** [comment]
Hi @ankeshanand .

Thanks for your detailed response! The HNS point makes a lot of sense!

I will re-try my baselines w/o the dropout, on more games. Will let you know if I have more issues.

Thanks again,
Kevin

**Comment by kevinghst** [comment]
Hi @ankeshanand 

I would like to re-open this issue. 

After running 10 seeds for all the 26 games using your latest commit (fixing default noisy std to paper value) and the commands from your README, I got the following results:

SPR, Mean HNS:              0.58          your paper: 0.70
SPR, Med HNS:                0.45.         your paper: 0.42
SPR no aug, Mean HNS:  0.46         your paper: 0.46
SPR no aug, Med HNS:    0.3            your paper: 0.3

The SPR no aug results are consistent with your paper. But the SPR Mean HS is 0.12 lower than the paper result, which to me seems pretty significant. 

Do you have any idea why that is? Have you confirmed that you were able to replicate the official numbers using the latest commit? I wonder if something is broken, or it's just noise.

Thanks,
Kevin

**Comment by MaxASchwarzer** [comment]
Hi Kevin,

I'm glad you were able to replicate our results!

As for the mean, though, I'm not that surprised.   I dug into the data in more detail a few weeks ago and found that the mean is dominated by one or two games with high scores and huge variance between seeds.  In fact, in SPR's original results, a single run on Up N Down added 0.05 to SPR's reported mean score, about half the difference between your result and ours!

Because of this, I think it's very hard to get a reliable estimate of mean performance without running substantially more than 10 seeds on those games.  My best guess from we've seen for SPR with different seeds or slightly different hyperparameters is that the true expected mean is probably somewhere around 0.62-0.64, but that's just informal.

Best,
Max

**Comment by kevinghst** [comment]
Got it, thanks!

**Comment by 0xJchen** [comment]
Hi, may I ask how do you run those games whose name is separated by a blank? For example, `KungFu Master`? I check the `atari-py` interface that the official name to create the game should be just 'KungFuMaster'. But when I invoke the script with `--game KungFuMaster or kungfumaster`, it says in function` game_path = atari_py.get_game_path(game)`, `Exception: ROM is missing for kungfumaster`. 
I could successfully run those games with names that are not separated by a blank, for example, `pong`. I make sure that I download the two atari rom files (ROM.rar, HC ROM.rar) and both installed them with `python -m atari_py.import_roms`. Could anyone please give me some suggestions on solving the above problem?

**Comment by ankeshanand** [comment]
You need to use lowercase and underscore for spaces, like `kung_fu_master`. See here for a list of all games: https://github.com/mila-iqia/spr/blob/release/src/rlpyt_utils.py#L19

**Comment by 0xJchen** [comment]
> You need to use lowercase and underscore for spaces, like `kung_fu_master`. See here for a list of all games: https://github.com/mila-iqia/spr/blob/release/src/rlpyt_utils.py#L19

Got it. Thanks for replying!

**Comment by yueyang130** [comment]
@kevinghst hi, i wanna to ask whether the numbers in the bracket in the first table you post is the standard deviation of 10 repeated experiements.

**Comment by kevinghst** [comment]
@yueyang130 yes that's correct

**Comment by yueyang130** [comment]
> Hi @ankeshanand
> 
> I would like to re-open this issue.
> 
> After running 10 seeds for all the 26 games using your latest commit (fixing default noisy std to paper value) and the commands from your README, I got the following results:
> 
> SPR, Mean HNS: 0.58 your paper: 0.70 SPR, Med HNS: 0.45. your paper: 0.42 SPR no aug, Mean HNS: 0.46 your paper: 0.46 SPR no aug, Med HNS: 0.3 your paper: 0.3
> 
> The SPR no aug results are consistent with your paper. But the SPR Mean HS is 0.12 lower than the paper result, which to me seems pretty significant.
> 
> Do you have any idea why that is? Have you confirmed that you were able to replicate the official numbers using the latest commit? I wonder if something is broken, or it's just noise.
> 
> Thanks, Kevin

Hi, kevinghst. I notice that you have re-run spr code. Actually, I'm doing that now, but I get stuck because the training takes way more time than 4.5 hours that the author says. My training on 12 GPUs, one A100 GPU and 30GB RAM takes about 8-10 hours. I'm trying to figure out why there is such a difference. I would be very appreciated if you can tell me your training sources and time and give me any advice.

**Comment by 0xJchen** [comment]
> > Hi @ankeshanand
> > I would like to re-open this issue.
> > After running 10 seeds for all the 26 games using your latest commit (fixing default noisy std to paper value) and the commands from your README, I got the following results:
> > SPR, Mean HNS: 0.58 your paper: 0.70 SPR, Med HNS: 0.45. your paper: 0.42 SPR no aug, Mean HNS: 0.46 your paper: 0.46 SPR no aug, Med HNS: 0.3 your paper: 0.3
> > The SPR no aug results are consistent with your paper. But the SPR Mean HS is 0.12 lower than the paper result, which to me seems pretty significant.
> > Do you have any idea why that is? Have you confirmed that you were able to replicate the official numbers using the latest commit? I wonder if something is broken, or it's just noise.
> > Thanks, Kevin
> 
> Hi, kevinghst. I notice that you have re-run spr code. Actually, I'm doing that now, but I get stuck because the training takes way more time than 4.5 hours that the author says. My training on 12 GPUs, one A100 GPU and 30GB RAM takes about 8-10 hours. I'm trying to figure out why there is such a difference. I would be very appreciated if you can tell me your training sources and time and give me any advice.

Can you report the CPU and GPU usage during training? I once met the same problem, and ending found that my GPU is not working thus it is simply trained on CPU, which would take roughly 10 hours.

**Comment by yueyang130** [comment]
> 



> > > Hi @ankeshanand
> > > I would like to re-open this issue.
> > > After running 10 seeds for all the 26 games using your latest commit (fixing default noisy std to paper value) and the commands from your README, I got the following results:
> > > SPR, Mean HNS: 0.58 your paper: 0.70 SPR, Med HNS: 0.45. your paper: 0.42 SPR no aug, Mean HNS: 0.46 your paper: 0.46 SPR no aug, Med HNS: 0.3 your paper: 0.3
> > > The SPR no aug results are consistent with your paper. But the SPR Mean HS is 0.12 lower than the paper result, which to me seems pretty significant.
> > > Do you have any idea why that is? Have you confirmed that you were able to replicate the official numbers using the latest commit? I wonder if something is broken, or it's just noise.
> > > Thanks, Kevin
> > 
> > 
> > Hi, kevinghst. I notice that you have re-run spr code. Actually, I'm doing that now, but I get stuck because the training takes way more time than 4.5 hours that the author says. My training on 12 GPUs, one A100 GPU and 30GB RAM takes about 8-10 hours. I'm trying to figure out why there is such a difference. I would be very appreciated if you can tell me your training sources and time and give me any advice.
> 
> Can you report the CPU and GPU usage during training? I once met the same problem, and ending found that my GPU is not working thus it is simply trained on CPU, which would take roughly 10 hours.

Yes,  the CPU is nearly fully used. And the utility of GPU is about 10%, which takes ~ 1.8GB GPU memory.  

Actually, I found backward is done on CPU (see [it](https://github.com/mila-iqia/spr/issues/28)), but the author says it has no influence on training. But when I move  the operations on GPU, the training time falls from nearly 20hours(Yeah, the original code takes nearly 20hours on 10 CPUs) to 8 hours. But it's still time-consuming. I doubt that I haven't moved all operations to GPU still.

Could you please tell me how you solve the problem. Thanks!

**Comment by 0xJchen** [comment]
>the original code takes nearly 20hours on 10 GPU

Do you mean CPU here? 

>But it's still time-consuming. I doubt that I haven't moved all operations to GPU still.

How about using a single GPU for training? In your case, do you find workload on all your GPUs (communication among GPUs might take a considerable amount of time when you got a lot of them) ?

**Comment by yueyang130** [comment]
> > the original code takes nearly 20hours on 10 GPU
> 
> Do you mean CPU here?
> 
> > But it's still time-consuming. I doubt that I haven't moved all operations to GPU still.
> 
> How about using a single GPU for training? In your case, do you find workload on all your GPUs?

Yes, sorry, I mean CPU.

I re-run origin spr code just now. It takes 2GB GPU memory and GPU-util is about 2%-5%. And I just use one GPU in all cases I mentioned above.

**Comment by 0xJchen** [comment]
>I re-run origin spr code just now. It takes 2GB GPU memory and GPU-util is about 2%-5%. And I just use one GPU in all cases I mentioned above.

So it still takes around 8 hours to finish?

Btw, do you see any warnings on incompatible CUDA or Torch versions?

**Comment by yueyang130** [comment]
> 

Yes, it still takes 8 or more hours to finish. And there is no warnings about CUDA and Torch version.
